File: app/src/main/java/de/tum/in/tumcampusapp/component/other/settings/SettingsFragment.java
package de.tum.in.tumcampusapp.component.other.settings;

import android.Manifest;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AlertDialog;
import androidx.preference.CheckBoxPreference;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceCategory;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import androidx.preference.SwitchPreferenceCompat;
import android.view.View;

import com.squareup.picasso.Picasso;

import java.io.IOException;
import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.api.tumonline.AccessTokenManager;
import de.tum.in.tumcampusapp.component.tumui.calendar.CalendarController;
import de.tum.in.tumcampusapp.component.ui.eduroam.SetupEduroamActivity;
import de.tum.in.tumcampusapp.component.ui.news.NewsController;
import de.tum.in.tumcampusapp.component.ui.news.model.NewsSources;
import de.tum.in.tumcampusapp.component.ui.onboarding.StartupActivity;
import de.tum.in.tumcampusapp.database.TcaDb;
import de.tum.in.tumcampusapp.service.BackgroundService;
import de.tum.in.tumcampusapp.service.SilenceService;
import de.tum.in.tumcampusapp.utils.Const;
import de.tum.in.tumcampusapp.utils.Utils;

public class SettingsFragment extends PreferenceFragmentCompat
        implements SharedPreferences.OnSharedPreferenceChangeListener, Preference.OnPreferenceClickListener {

    public static final String FRAGMENT_TAG = "my_preference_fragment";
    private static final String BUTTON_LOGOUT = "button_logout";
    private static final String SETUP_EDUROAM = "card_eduroam_setup";

    private FragmentActivity mContext;

    @Override
    public void onCreatePreferences(Bundle bundle, String rootKey) {
        // Load the correct preference category
        setPreferencesFromResource(R.xml.settings, rootKey);
        mContext = getActivity();

        populateNewsSources();
        setUpEmployeeSettings();

        // Disables silence service if the app is used without TUMOnline access
        SwitchPreferenceCompat silentSwitch =
                (SwitchPreferenceCompat) findPreference(Const.SILENCE_SERVICE);
        if (silentSwitch != null && !AccessTokenManager.hasValidAccessToken(getContext())) {
            silentSwitch.setEnabled(false);
        }

        // Only do these things if we are in the root of the preferences
        if (rootKey == null) {
            // Click listener for preference list entries. Used to simulate a button
            // (since it is not possible to add a button to the preferences screen)
            findPreference(BUTTON_LOGOUT).setOnPreferenceClickListener(this);

            setSummary("card_default_campus");
            setSummary("silent_mode_set_to");
            setSummary("background_mode_set_to");
        } else if (rootKey.equals("card_cafeteria")) {
            setSummary("card_cafeteria_default_G");
            setSummary("card_cafeteria_default_K");
            setSummary("card_cafeteria_default_W");
            setSummary("card_role");
        } else if (rootKey.equals("card_mvv")) {
            setSummary("card_stations_default_G");
            setSummary("card_stations_default_C");
            setSummary("card_stations_default_K");
        } else if (rootKey.equals("card_eduroam")) {
            findPreference(SETUP_EDUROAM).setOnPreferenceClickListener(this);
        }

        // Register the change listener to react immediately on changes
        PreferenceManager.getDefaultSharedPreferences(mContext)
                         .registerOnSharedPreferenceChangeListener(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        // Set the default white background in the view so as to avoid transparency
        view.setBackgroundColor(Color.WHITE);
    }

    private void populateNewsSources() {
        PreferenceCategory newsSourcesPreference =
                (PreferenceCategory) findPreference("card_news_sources");

        NewsController newsController = new NewsController(mContext);
        List<NewsSources> newsSources = newsController.getNewsSources();
        for (NewsSources newsSource : newsSources) {
            final CheckBoxPreference pref = new CheckBoxPreference(mContext);
            pref.setKey("card_news_source_" + newsSource.getId());
            pref.setDefaultValue(true);
            
            // reserve space so that when the icon is loaded the text is not moved again
            pref.setIconSpaceReserved(true);

            // Load news source icon in background and set it
            final String url = newsSource.getIcon();
            if (!url.trim().isEmpty()) { // Skip News that do not have a image
                new Thread(() -> {
                    try {
                        Bitmap bmp = Picasso.get().load(url).get();
                        mContext.runOnUiThread(() -> {
                            if(isAdded()){
                                pref.setIcon(new BitmapDrawable(getResources(), bmp));
                            }
                        });
                    } catch (IOException e) {
                        // ignore
                    }
                }).start();
            }

            pref.setTitle(newsSource.getTitle());
            if (newsSourcesPreference != null) {
                newsSourcesPreference.addPreference(pref);
            }
        }
    }

    /**
     * Disable setting for non-employees.
     */
    private void setUpEmployeeSettings() {
        boolean isEmployee = !Utils.getSetting(mContext, Const.TUMO_EMPLOYEE_ID, "").isEmpty();
        if (!isEmployee) {
            findPreference(Const.EMPLOYEE_MODE).setEnabled(false);
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        Preference pref = findPreference(key);
        if (pref instanceof ListPreference) {
            ListPreference listPreference = (ListPreference) pref;
            String entry = listPreference.getEntry().toString();
            listPreference.setSummary(entry);
        }

        // When newspread selection changes
        // deselect all newspread sources and select only the
        // selected source if one of all was selected before
        if ("news_newspread".equals(key)) {
            SharedPreferences.Editor e = sharedPreferences.edit();
            boolean value = false;
            for (int i = 7; i < 14; i++) {
                if (sharedPreferences.getBoolean("card_news_source_" + i, false)) {
                    value = true;
                }
                e.putBoolean("card_news_source_" + i, false);
            }
            String newSource = sharedPreferences.getString(key, "7");
            e.putBoolean("card_news_source_" + newSource, value);
            e.apply();
        }

        // If the silent mode was activated, start the service. This will invoke
        // the service to call onHandleIntent which checks available lectures
        if (key.equals(Const.SILENCE_SERVICE)) {
            Intent service = new Intent(mContext, SilenceService.class);
            if (sharedPreferences.getBoolean(key, false)) {
                if (!SilenceService.hasPermissions(mContext)) {
                    // disable until silence service permission is resolved
                    SwitchPreferenceCompat silenceSwitch =
                            (SwitchPreferenceCompat) findPreference(Const.SILENCE_SERVICE);
                    if (silenceSwitch != null) {
                        silenceSwitch.setChecked(false);
                    }
                    Utils.setSetting(mContext, Const.SILENCE_SERVICE, false);

                    SilenceService.requestPermissions(mContext);
                } else {
                    mContext.startService(service);
                }
            } else {
                mContext.stopService(service);
            }
        }

        // If the background mode was activated, start the service. This will invoke
        // the service to call onHandleIntent which updates all background data
        if (key.equals(Const.BACKGROUND_MODE)) {
            Intent service = new Intent(mContext, BackgroundService.class);
            if (sharedPreferences.getBoolean(key, false)) {
                mContext.startService(service);
            } else {
                mContext.stopService(service);
            }
        }
    }

    private void setSummary(CharSequence key) {
        Preference pref = findPreference(key);
        if (pref instanceof ListPreference) {
            ListPreference listPref = (ListPreference) pref;
            String entry = listPref.getEntry().toString();
            listPref.setSummary(entry);
        }
    }

    /**
     * Handle all clicks on 'button'-preferences
     *
     * @param preference Preference that has been clicked
     * @return True, if handled
     */
    @Override
    public boolean onPreferenceClick(Preference preference) {
        final String key = preference.getKey();

        switch (key) {
            case SETUP_EDUROAM:
                startActivity(new Intent(getContext(), SetupEduroamActivity.class));
                break;
            case BUTTON_LOGOUT:
                showLogoutDialog();
                break;
            default:
                return false;
        }

        return true;
    }

    private void showLogoutDialog() {
        AlertDialog dialog = new AlertDialog.Builder(mContext)
                .setMessage(R.string.logout_message)
                .setPositiveButton(R.string.logout, ((dialogInterface, i) -> logout()))
                .setNegativeButton(R.string.cancel, null)
                .create();

        if (dialog.getWindow() != null) {
            dialog.getWindow().setBackgroundDrawableResource(R.drawable.rounded_corners_background);
        }

        dialog.show();
    }

    private void logout() {
        clearData();
        startActivity(new Intent(mContext, StartupActivity.class));
        mContext.finish();
    }

    private void clearData() {
        TcaDb.resetDb(mContext);

        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(mContext);
        sharedPrefs.edit().clear().apply();

        // Remove all notifications that are currently shown
        NotificationManager notificationManager =
                (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager != null) {
            notificationManager.cancelAll();
        }

        int readCalendar = ContextCompat.checkSelfPermission(
                requireActivity(), Manifest.permission.READ_CALENDAR);
        int writeCalendar = ContextCompat.checkSelfPermission(
                requireActivity(), Manifest.permission.WRITE_CALENDAR);

        // Delete local calendar
        Utils.setSetting(mContext, Const.SYNC_CALENDAR, false);
        if (readCalendar == PackageManager.PERMISSION_GRANTED &&
                writeCalendar == PackageManager.PERMISSION_GRANTED) {
            CalendarController.deleteLocalCalendar(mContext);
        }
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/other/settings/SettingsFragment.kt
package de.tum.`in`.tumcampusapp.component.other.settings

import android.Manifest.permission.READ_CALENDAR
import android.Manifest.permission.WRITE_CALENDAR
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat.checkSelfPermission
import androidx.core.content.edit
import androidx.core.os.bundleOf
import androidx.preference.CheckBoxPreference
import androidx.preference.ListPreference
import androidx.preference.MultiSelectListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.SwitchPreferenceCompat
import com.squareup.picasso.Picasso
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.api.tumonline.AccessTokenManager
import de.tum.`in`.tumcampusapp.component.tumui.calendar.CalendarController
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.repository.CafeteriaLocalRepository
import de.tum.`in`.tumcampusapp.component.ui.eduroam.SetupEduroamActivity
import de.tum.`in`.tumcampusapp.component.ui.news.NewsController
import de.tum.`in`.tumcampusapp.component.ui.onboarding.StartupActivity
import de.tum.`in`.tumcampusapp.database.TcaDb
import de.tum.`in`.tumcampusapp.di.injector
import de.tum.`in`.tumcampusapp.service.SilenceService
import de.tum.`in`.tumcampusapp.service.StartSyncReceiver
import de.tum.`in`.tumcampusapp.utils.Const
import de.tum.`in`.tumcampusapp.utils.Utils
import de.tum.`in`.tumcampusapp.utils.plusAssign
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import org.jetbrains.anko.defaultSharedPreferences
import org.jetbrains.anko.notificationManager
import java.util.concurrent.ExecutionException
import javax.inject.Inject

class SettingsFragment : PreferenceFragmentCompat(), Preference.OnPreferenceClickListener,
        SharedPreferences.OnSharedPreferenceChangeListener {

    private val themeProvider by lazy { ThemeProvider(requireContext()) }

    private val compositeDisposable = CompositeDisposable()

    @Inject
    lateinit var cafeteriaLocalRepository: CafeteriaLocalRepository

    @Inject
    lateinit var newsController: NewsController

    override fun onAttach(context: Context) {
        super.onAttach(context)
        injector.inject(this)
    }

    override fun onCreatePreferences(
        savedInstanceState: Bundle?,
        rootKey: String?
    ) {
        setPreferencesFromResource(R.xml.settings, rootKey)
        populateNewsSources()
        setUpEmployeeSettings()

        // Disables silence service and logout if the app is used without TUMOnline access
        val silentSwitch = findPreference(Const.SILENCE_SERVICE) as? SwitchPreferenceCompat
        val logoutButton = findPreference(BUTTON_LOGOUT)
        if (!AccessTokenManager.hasValidAccessToken(context)) {
            if (silentSwitch != null) {
                silentSwitch.isEnabled = false
            }
            logoutButton?.isVisible = false
        }

        // Only do these things if we are in the root of the preferences
        if (rootKey == null) {
            // Click listener for preference list entries. Used to simulate a button
            // (since it is not possible to add a button to the preferences screen)
            findPreference(BUTTON_LOGOUT).onPreferenceClickListener = this
            setSummary("language_preference")
            setSummary(Const.DESIGN_THEME)
            setSummary("card_default_campus")
            setSummary("silent_mode_set_to")
            setSummary("background_mode_set_to")
        } else if (rootKey == "card_cafeteria") {
            setSummary("card_cafeteria_default_G")
            setSummary("card_cafeteria_default_K")
            setSummary("card_cafeteria_default_W")
            setSummary("card_role")
            initCafeteriaCardSelections()
        } else if (rootKey == "card_mvv") {
            setSummary("card_stations_default_G")
            setSummary("card_stations_default_C")
            setSummary("card_stations_default_K")
        } else if (rootKey == "card_eduroam") {
            findPreference(SETUP_EDUROAM).onPreferenceClickListener = this
        }

        requireContext().defaultSharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    private fun populateNewsSources() {
        val newsSourcesPreference = findPreference("card_news_sources") as? PreferenceCategory ?: return
        val newsSources = newsController.newsSources

        for (newsSource in newsSources) {
            val pref = CheckBoxPreference(requireContext())
            pref.key = "card_news_source_" + newsSource.id
            pref.setDefaultValue(true)

            // Reserve space so that when the icon is loaded the text is not moved again
            pref.isIconSpaceReserved = true

            // Load news source icon in background and set it
            val url = newsSource.icon
            if (url.isNotBlank()) {
                loadNewsSourceIcon(pref, url)
            }

            pref.title = newsSource.title
            newsSourcesPreference.addPreference(pref)
        }
    }

    private fun loadNewsSourceIcon(
        preference: Preference,
        url: String
    ) {
        compositeDisposable += Single
                .fromCallable { Picasso.get().load(url).get() }
                .subscribeOn(Schedulers.io())
                .map { BitmapDrawable(resources, it) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(preference::setIcon, Utils::log)
    }

    /**
     * Disable setting for non-employees.
     */
    private fun setUpEmployeeSettings() {
        val isEmployee = Utils.getSetting(requireContext(), Const.TUMO_EMPLOYEE_ID, "").isNotEmpty()
        val checkbox = findPreference(Const.EMPLOYEE_MODE) ?: return
        if (!isEmployee) {
            checkbox.isEnabled = false
        }
    }

    override fun onSharedPreferenceChanged(
        sharedPrefs: SharedPreferences,
        key: String?
    ) {
        if (key == null) {
            return
        }
        setSummary(key)

        // When newspread selection changes
        // deselect all newspread sources and select only the
        // selected source if one of all was selected before
        if (key == "news_newspread") {
            val newspreadIds = 7..13
            val value = newspreadIds.any { sharedPrefs.getBoolean("card_news_source_$it", false) }
            val newSource = sharedPrefs.getString(key, "7")

            sharedPrefs.edit {
                newspreadIds.forEach { putBoolean("card_news_source_$it", false) }
                putBoolean("card_news_source_$newSource", value)
            }
        }

        // If the silent mode was activated, start the service. This will invoke
        // the service to call onHandleIntent which checks available lectures
        if (key == Const.SILENCE_SERVICE) {
            val service = Intent(requireContext(), SilenceService::class.java)
            if (sharedPrefs.getBoolean(key, false)) {
                if (!SilenceService.hasPermissions(requireContext())) {
                    // disable until silence service permission is resolved
                    val silenceSwitch = findPreference(Const.SILENCE_SERVICE) as SwitchPreferenceCompat
                    silenceSwitch.isChecked = false
                    Utils.setSetting(requireContext(), Const.SILENCE_SERVICE, false)
                    SilenceService.requestPermissions(requireContext())
                } else {
                    requireContext().startService(service)
                }
            } else {
                requireContext().stopService(service)
            }
        }

        // Change design theme
        if (key == Const.DESIGN_THEME) {
            val theme = themeProvider.getTheme(sharedPrefs.getString(key, "system")!!)
            AppCompatDelegate.setDefaultNightMode(theme)
        }

        // If the background mode was activated, start the service. This will invoke
        // the service to call onHandleIntent which updates all background data
        if (key == Const.BACKGROUND_MODE) {
            if (sharedPrefs.getBoolean(key, false)) {
                StartSyncReceiver.startBackground(requireContext())
            } else {
                StartSyncReceiver.cancelBackground()
            }
        }

        // restart app after language change
        if (key == "language_preference" && activity != null) {
            (activity as SettingsActivity).restartApp()
        }
    }

    private fun initCafeteriaCardSelections() {
        val cafeterias = cafeteriaLocalRepository
                .getAllCafeterias()
                .blockingFirst()
                .sortedBy { it.name }

        val cafeteriaByLocationName = getString(R.string.settings_cafeteria_depending_on_location)
        val cafeteriaNames = listOf(cafeteriaByLocationName) + cafeterias.map { it.name }

        val cafeteriaByLocationId = Const.CAFETERIA_BY_LOCATION_SETTINGS_ID
        val cafeteriaIds = listOf(cafeteriaByLocationId) + cafeterias.map { it.id.toString() }

        val preference = findPreference(Const.CAFETERIA_CARDS_SETTING) as MultiSelectListPreference
        preference.entries = cafeteriaNames.toTypedArray()
        preference.entryValues = cafeteriaIds.toTypedArray()

        preference.setOnPreferenceChangeListener { pref, newValue ->
            (pref as MultiSelectListPreference).values = newValue as Set<String>
            setCafeteriaCardsSummary(preference)
            false
        }

        setCafeteriaCardsSummary(preference)
    }

    private fun setSummary(key: CharSequence) {
        val pref = findPreference(key)
        if (pref is ListPreference) {
            val entry = pref.entry.toString()
            pref.summary = entry
        }
    }

    private fun setCafeteriaCardsSummary(preference: MultiSelectListPreference) {
        val values = preference.values
        if (values.isEmpty()) {
            preference.setSummary(R.string.settings_no_location_selected)
        } else {
            preference.summary = values
                    .map { preference.findIndexOfValue(it) }
                    .map { preference.entries[it] }
                    .map { it.toString() }
                    .sorted()
                    .joinToString(", ")
        }
    }

    override fun onPreferenceClick(
        preference: Preference?
    ): Boolean {
        when (preference?.key) {
            SETUP_EDUROAM -> startActivity(Intent(context, SetupEduroamActivity::class.java))
            BUTTON_LOGOUT -> showLogoutDialog(R.string.logout_title, R.string.logout_message)
            else -> return false
        }
        return true
    }

    private fun showLogoutDialog(title: Int, message: Int) {
        AlertDialog.Builder(requireContext())
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(R.string.logout) { _, _ -> logout() }
                .setNegativeButton(R.string.cancel, null)
                .create()
                .apply {
                    window?.setBackgroundDrawableResource(R.drawable.rounded_corners_background)
                }
                .show()
    }

    private fun logout() {
        try {
            clearData()
        } catch (e: Exception) {
            Utils.log(e)
            showLogoutDialog(R.string.logout_error_title, R.string.logout_try_again)
            return
        }

        startActivity(Intent(requireContext(), StartupActivity::class.java))
        requireActivity().finish()
    }

    @SuppressLint("ApplySharedPref")
    @Throws(ExecutionException::class, InterruptedException::class)
    private fun clearData() {
        TcaDb.resetDb(requireContext())
        requireContext().defaultSharedPreferences.edit().clear().commit()

        // Remove all notifications that are currently shown
        requireContext().notificationManager.cancelAll()

        val readCalendar = checkSelfPermission(requireActivity(), READ_CALENDAR)
        val writeCalendar = checkSelfPermission(requireActivity(), WRITE_CALENDAR)

        // Delete local calendar
        Utils.setSetting(requireContext(), Const.SYNC_CALENDAR, false)
        if (readCalendar == PERMISSION_GRANTED && writeCalendar == PERMISSION_GRANTED) {
            CalendarController.deleteLocalCalendar(requireContext())
        }
    }

    override fun onDestroy() {
        compositeDisposable.dispose()
        super.onDestroy()
    }

    companion object {

        private const val BUTTON_LOGOUT = "button_logout"
        private const val SETUP_EDUROAM = "card_eduroam_setup"

        fun newInstance(
            key: String?
        ) = SettingsFragment().apply { arguments = bundleOf(ARG_PREFERENCE_ROOT to key) }
    }
}
